common/text/tree_utils.h

```diff
@@ -67,36 +67,49 @@ const SyntaxTreeLeaf& SymbolCastToLeaf(const Symbol&);
 // multiple children.
 const Symbol* DescendThroughSingletons(const Symbol& symbol);
 
-// Succeeds if node's enum matches 'node_enum'.
+// Succeeds and returns node if node's enum matches 'node_enum'.
 // Returns same node reference, so that anywhere that expects a SyntaxTreeNode
-// can be passed CheckNodeEnum(node, node_enum).
+// can be passed MatchNodeEnumOrNull(node, node_enum).
 template <typename E>
-const SyntaxTreeNode& CheckNodeEnum(const SyntaxTreeNode& node,
-                                    E expected_node_enum) {
+const SyntaxTreeNode* MatchNodeEnumOrNull(const SyntaxTreeNode& node,
+                                          E expected_node_enum) {
   // Uses operator<<(std::ostream&, E) for diagnostics.
-  CHECK_EQ(E(node.Tag().tag), expected_node_enum);
-  return node;
+  const bool enum_matches = (E(node.Tag().tag) == expected_node_enum);
+  if (!enum_matches) {
+    LOG(ERROR) << "Node: Programming error: expected " << expected_node_enum
```

Comment:
Would it be possible to add more information to VLOG? The whole tree etc. We already got tags from check-failures and they weren't useful.